src/actor/Verifier.ts

```diff
@@ -0,0 +1,148 @@
+import PublicAttesterIdentity from '../attesteridentity/PublicAttesterIdentity'
+
+/**
+ * A helper class to initiate a verification by creating a presentation request which is built
+ * on a specific [[CType]] and attributes of the [[Claim]] the verifier requires to see.
+ */
+export class PresentationRequestBuilder {
+  private builder: gabi.CombinedRequestBuilder
+  private ctypes: Array<CType['hash']>
+  constructor() {
+    this.builder = new gabi.CombinedRequestBuilder()
+    this.ctypes = []
+  }
+
+  /**
+   * Initiates a verification by creating a presentation request for a specific [[CType]].
+   * Note that you are required to call [[finalize]] on the request to conclude it.
+   *
+   * @param p The parameter object.
+   * @param p.ctypehash The SHA-256 hash of the [[CType]].
+   * @param p.attributes A list of attributes of the [[Credential]]s the Verifier has to see in order to verify it.
+   * @param p.legitimations An optional boolean representing whether the Verifier requests to see the legitimations of the Attesters which signed the [[Credential]]s.
+   * @param p.delegationId An optional boolean representing whether the Verifier requests to see the Attesters' unique delegation identifiers.
+   * @param p.reqUpdatedAfter The optional minimum required timestamp on which the [[Credential]] needs to be updated.
+   * The default value for this is the current date.
+   * @returns A [[PresentationRequestBuilder]] on which you need to call [[finalize]] to complete the presentation request.
+   */
+  public requestPresentationForCtype({
+    ctypeHash,
+    attributes,
+    legitimations,
+    delegationId,
+    reqUpdatedAfter = new Date(),
+  }: {
+    attributes: string[]
+    ctypeHash: CType['hash']
+    legitimations?: boolean
+    delegationId?: boolean
+    reqUpdatedAfter?: Date
+  }): PresentationRequestBuilder {
+    const rawAttribute = attributes.map(attr => `claim.contents.${attr}`)
+    rawAttribute.push('claim.cTypeHash')
+    if (typeof legitimations !== 'undefined' && legitimations) {
+      rawAttribute.push('legitimation')
+    }
+    if (typeof delegationId !== 'undefined' && delegationId) {
+      rawAttribute.push('delegationId')
+    }
+    this.builder.requestPresentation({
+      requestedAttributes: rawAttribute,
+      reqUpdatedAfter,
+    })
+    this.ctypes.push(ctypeHash)
+    return this
+  }
+
+  /**
+   * [ASYNC] Concludes the presentation request.
+   *
+   * @param allowPE A boolean representing whether the verifier accepts a privacy enhanced presentation.
+   * @returns A session and a message object.
+   * The **session** object will be used in [[verifyPresentation]] and should be kept private by the verifier.
+   * The **message** object should be sent to the Claimer and used in [[createPresentation]].
+   */
+  public async finalize(
+    allowPE: boolean
+  ): Promise<[gabi.CombinedVerificationSession, IRequestClaimsForCTypes]> {
+    const { session, message } = await this.builder.finalise()
```

Comment:
Is it necessary to call go / go-wasm, when I only want a classic kilt credential?

Reply:
It is less code that way. We would need to create a public only request. But This would only add complexity since we still need to be able to create a public response from a privacy enhanced request. There are three different cases for this

1. Verifier allows public and pe -> claimer responds with pe
1. Verifier allows public and pe -> claimer responds with public
1. Verifier only allows public -> claimer responds with public

At the moment these are all the same request with just a single flag signaling if we would allow PE responses. If we would add a public request we would only improve case 3. In case 2 we would either need to send both requests or derive the public request from the private request.

Sending both requests increases the message size significantly. But if we derive the public request from the pe request, why do we need a special public request in the first place?

I would only fix this if the wasm call causes any performance issues. Until then we maybe already changed to another credential systems (one written in rust ❤️ 🥥 ).